stdlib/public/core/ObjectIdentifier.swift

```diff
@@ -14,6 +14,11 @@
 ///
 /// In Swift, only class instances and metatypes have unique identities. There
 /// is no notion of identity for structs, enums, functions, or tuples.
+///
+/// `ObjectIdentifier` is only guaranteed to remain unique for the
+/// lifetime of an object. When the instance gets deallocated, its object 
+/// identifier may be reused for a different object. (Internally, objects are
+/// identified by their memory location.)
```

Comment:
I'm not really sure whether we really want to have wording for allocation/deallocation/memory locations here. That makes it seem like we are making some commitments about the memory allocation under the hood, but I'm not sure if we want to do that. How about something simpler:

```
The unique identity for a class instance is only valid for comparisons during
the lifetime of the instance. The unique identity for a metatype is always valid.
```

You could also omit the second sentence I suggested above if you think it's not helpful.